// firebase-function/lib/src/logger/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const StringBuilder_1 = require("../StringBuilder");
const LogLevel_1 = require("./LogLevel");
const flatted_1 = require("flatted");
class Logger {
    verbose;
    properties;
    currentIndent;
    constructor(verbose, properties, currentIndent = 0) {
        this.verbose = verbose;
        this.properties = properties;
        this.currentIndent = currentIndent;
    }
    static start(verbose, functionName, details, logLevel = 'debug') {
        const property = {
            functionName: functionName,
            logLevel: logLevel,
            indent: 0,
            details: details
        };
        return new Logger(verbose, [property]);
    }
    get nextIndent() {
        return new Logger(this.verbose, this.properties, this.currentIndent + 1);
    }
    log(functionName, details, logLevel = 'debug') {
        this.properties.push({
            functionName: functionName,
            logLevel: logLevel,
            indent: this.currentIndent,
            details: details
        });
    }
    propertyString(property) {
        const builder = new StringBuilder_1.StringBuilder();
        builder.appendLine(`${' '.repeat(2 * property.indent)}| ${LogLevel_1.LogLevel.coloredText(property.logLevel, `[${property.functionName}]`, this.verbose === 'colored' || this.verbose === 'coloredVerbose')}`);
        if (property.details !== undefined && (this.verbose === 'verbose' || this.verbose === 'coloredVerbose')) {
            for (const entry of Object.entries(property.details))
                builder.append(this.detailString(property.indent, entry[0], entry[1]));
        }
        return builder.toString();
    }
    detailString(indent, key, detail) {
        const builder = new StringBuilder_1.StringBuilder();
        const jsonLines = (0, flatted_1.stringify)(detail, undefined, '  ')?.split('\n') ?? [''];
        builder.appendLine(`${' '.repeat(2 * indent)}| ${`${key}: \x1b[40m\x1b[2m${jsonLines.shift() ?? ''}\x1b[0m`}`);
        for (const line of jsonLines)
            builder.appendLine(`${' '.repeat(2 * indent)}| ${' '.repeat(key.length + 2)}\x1b[40m\x1b[2m${line}\x1b[0m`);
        return builder.toString();
    }
    get completeLog() {
        const builder = new StringBuilder_1.StringBuilder();
        for (const property of this.properties)
            builder.append(this.propertyString(property));
        return builder.toString();
    }
}
exports.Logger = Logger;
